2695 added blog and author details (#2712)
opensearch-project/project-website: _posts/2024-04-02-Multilingual-Search.md

---
layout: post
title:  "Multilingual search with OpenSearch"
authors:
- roigam

> date: 2024-04-01 13:30:00 -0600

Reviewer: Please update the publish date to 4/25/2024

Author: done

categories:
  - technical-post
meta_keywords: semantic search engine, multilingual search, keyword and natural language search, search relevance, embeddings
meta_description: Learn how to use OpenSearch to conduct multilingual search with models that support multilingual embeddings.

> ---

Reviewer: Please update the meta with the following:

**meta_keywords:** OpenSearch semantic search, multilingual search, multilingual embedding models, natural language and search

**meta_description:**  Learn how OpenSearch semantic search can help you understand the intent of a search query and conduct multilingual search with models that support multilingual embeddings.

Author: done

* What happens if the catalog is in one language, and my users search in other languages?
* What if my catalog items are defined in different languages?
* How can I handle search requests or catalog items in languages that are not supported out-of-the-box in [OpenSearch language analyzers](https://opensearch.org/docs/latest/analyzers/language-analyzers/) (e.g., Hebrew)? 

Some of the above challenges could be solved using schema changes, custom code and ETL pipelines, but would greatly complicate the code base, require maintenance and labor time.
In this blog post we discuss how can we use semantic search with multilingual embedding models to overcome the challenges above. 

## Multilingual models
Transforming text into a vector is in the heart of semantic search. 
Embedding models are trained on vast amounts of data and are able to convert text into vectors in a way that capture inherent properties and relationships in real-world data.
Some embedding models are trained on more than one language and therefore can be used to search within a language or across languages. 

> Examples for such models are [Cohere’s multilingual embedding model](https://docs.cohere.com/docs/multilingual-language-models) and [Amazon Titan Embeddings G1 - Text G1](https://docs.aws.amazon.com/bedrock/latest/userguide/titan-embedding-models.html)

Reviewer 2: ```suggestion
Examples of such models are [Cohere’s multilingual embedding model](https://docs.cohere.com/docs/multilingual-language-models) and [Amazon Titan Embeddings G1 - Text G1](https://docs.aws.amazon.com/bedrock/latest/userguide/titan-embedding-models.html).
```

> OpenSearch allows you to integrate multilingual embedding models into your search workflow with the following approaches:

Reviewer 2: ```suggestion
OpenSearch allows you to integrate multilingual embedding models into your search workflow using the following approaches:
```

* Local models that are uploaded to your cluster locally, such as `huggingface/sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`

> * Externally hosted models on third-party platform such as [Cohere](https://cohere.com/) and [Amazon Bedrock](https://aws.amazon.com/bedrock/).

Reviewer 2: ```suggestion
* Externally hosted models on third-party platforms such as [Cohere](https://cohere.com/) or [Amazon Bedrock](https://aws.amazon.com/bedrock/).
```

> For an explanation on the tradeoffs between the approaches see [choosing a model](https://opensearch.org/docs/latest/ml-commons-plugin/integrating-ml-models/#choosing-a-model)

Reviewer 2: ```suggestion
For an explanation of the tradeoffs between these approaches, see [Choosing a model](https://opensearch.org/docs/latest/ml-commons-plugin/integrating-ml-models/#choosing-a-model).
```

## Using multilingual embedding models in OpenSearch

> We will now go over the steps required to allow multilingual search using an externally hosted model on Cohere. 

Reviewer 2: ```suggestion
We will now go over the steps required to allow multilingual search using a model externally hosted on Cohere. 
```

> For a tutorial showing how to use a local model please follow [this guide](https://repost.aws/articles/ARwaKLdzGDShq-O2OMU2Nnhg/multilingual-search-with-amazon-opensearch).

Reviewer 2: ```suggestion
For a tutorial showing how to use a local model, please see [this guide](https://repost.aws/articles/ARwaKLdzGDShq-O2OMU2Nnhg/multilingual-search-with-amazon-opensearch).
```

> Connecting to externally hosted models is done through [Connectors](https://opensearch.org/docs/latest/ml-commons-plugin/remote-models/connectors/). 

Reviewer 2: ```suggestion
You can connect to externally hosted models through [connectors](https://opensearch.org/docs/latest/ml-commons-plugin/remote-models/connectors/). 
```

        "field_map": {
          "description": "description_vector"
        }
      }
    }
  ]
}
```

### Creating an index
We now create a k-NN index with the pipeline created in the previous step. Note the `dimension` for `description_vector` must match the model dimension you use.
```json
PUT /cohere_multilingual_english_catalog
{
  "settings": {
      "index": {
        "number_of_shards": 1,
        "number_of_replicas": 1
      },
      "index.knn": true,
      "index.knn.space_type": "cosinesimil",
      "default_pipeline": "cohere_multilingual_pipeline"
  },
  "mappings": {
      "properties": {
          "description_vector": {
              "type": "knn_vector",
              "dimension": 1024,
              "method": {
                  "name": "hnsw",
                  "space_type": "l2",
                  "engine": "faiss"
              },
              "store": true
          },
          "description": {
              "type": "text",
              "store": true
          }
      }
  }
}
```

More information on creating a k-NN index can be found [here](https://opensearch.org/docs/latest/search-plugins/knn/knn-index/).

### Ingesting data 

> We are ready to ingest our catalog. In the example below we will use the `_bulk` API, and show just a sub set of the data. Note that the full corpus of data holds 2000 documents.

Reviewer 2: ```suggestion
You are now ready to ingest your catalog. The following example uses the `_bulk` API and shows only a subset of the data. Note that the full data corpus holds 2,000 documents.
```

{"index": {"_index": "cohere_multilingual_english_catalog", "_id": 8}}
{"gender": "Men", "masterCategory": "Accessories", "subCategory": "Socks", "articleType": "Socks", "baseColour": "Navy Blue", "season": "Summer", "year": "2012", "usage": "Casual", "productDisplayName": "Puma Men Pack of 3 Socks", "description": "Men Accessories Socks Socks Navy Blue Summer 2012 Casual Puma Men Pack of 3 Socks"}
{"index": {"_index": "cohere_multilingual_english_catalog", "_id": 9}}
{"gender": "Men", "masterCategory": "Accessories", "subCategory": "Watches", "articleType": "Watches", "baseColour": "Black", "season": "Winter", "year": "2016", "usage": "Casual", "productDisplayName": "Skagen Men Black Watch", "description": "Men Accessories Watches Watches Black Winter 2016 Casual Skagen Men Black Watch"}
...
```
Once the data is ingested, the `description` field is converted into text embeddings and the embeddings are stored in `description_vector` using the ingestion pipeline and connector we created in earlier stages.

### Performing multilingual search requests

> We can now search our catalog. Semantic search can be done in English, or all languages the [model supports](https://docs.cohere.com/docs/supported-languages).  

Reviewer 2: ```suggestion
You can now search your catalog. Semantic search can be performed in English or in any of the model's [supported languages](https://docs.cohere.com/docs/supported-languages).  
```

#### Searching in English

> We can still run a semantic search in the English.

Reviewer 2: ```suggestion
The following is an example semantic search in English:
```

        "_source": {
          "subCategory": "Fragrance",
          "description": "Men Personal Care Fragrance Deodorant Blue Spring 2017 Casual Rasasi Men Deserve Deo"
        }
      },
      {
        ...
        "_score": 0.5110825,
        "_source": {
          "subCategory": "Jewellery",
          "description": "Men Accessories Jewellery Ring Steel Summer 2013 Casual Revv Men Steel Ring"
        }
      },
      {
        ...
        "_score": 0.5016889,
        "_source": {
          "subCategory": "Sandal",
          "description": "Men Footwear Sandal Sandals Olive Summer 2012 Casual Bata Men Hummer Grey Sandals"
        }
      },
      {
        ...
        "_score": 0.50156057,
        "_source": {
          "subCategory": "Fragrance",
          "description": "Men Personal Care Fragrance Fragrance Gift Set Red Spring 2017 Casual Reebok Men Reeplay Perfume & Deodorant Set"
        }
      },
      {
        ...
        "_score": 0.50010604,
        "_source": {
          "subCategory": "Sandal",
          "description": "Men Footwear Sandal Sandals Brown Summer 2013 Casual Gas Men Aventura Sandal"
        }
      }
    ]
}
```
#### Searching in other languages

> Since the embedding was done using a multilingual embedding model, we can search in any of the languages supported by said model. 

Reviewer 2: ```suggestion
Because the embedding was performed using a multilingual embedding model, you can search in any of the languages supported by the model. 
```

> As an example, this includes Hebrew that is not supported out-of-the-box in OpenSearch:

Reviewer 2: ```suggestion
As an example, this includes Hebrew, which is not supported out-of-the-box in OpenSearch:
```

    "hits": [
      {
        ...
        "_score": 0.62954986,
        "_source": {
          "subCategory": "Sandal",
          "description": "Men Footwear Sandal Sandals Black Summer 2012 Casual Coolers Men Black Sandals"
        }
      },
      {
       ...
        "_score": 0.62954986,
        "_source": {
          "subCategory": "Sandal",
          "description": "Men Footwear Sandal Sandals Black Summer 2012 Casual Coolers Men Black Sandals"
        }
      },
      {
        ...
        "_score": 0.62518847,
        "_source": {
          "subCategory": "Shoes",
          "description": "Men Footwear Shoes Sports Shoes Black Summer 2012 Casual Nike Men Air Twilight Black Shoes"
        }
      },
      {
        ...
        "_score": 0.62431824,
        "_source": {
          "subCategory": "Shoes",
          "description": "Men Footwear Shoes Sports Shoes Black Summer 2011 Sports Nike Men's Elite Black Shoe"
        }
      },
      {
        ...
        "_score": 0.6242107,
        "_source": {
          "subCategory": "Sandal",
          "description": "Men Footwear Sandal Sandals Olive Summer 2012 Casual Bata Men Hummer Grey Sandals"
        }
      }
    ]
}
```

> The same approach as above can be used if the catalog itself was in Hebrew, or the items in the catalog where in different languages. 

Reviewer 2: ```suggestion
The same approach could be used if the catalog itself were in Hebrew or if the items in the catalog were described in different languages. 
```

> Important to note that lexical search can still be done on the index, and OpenSearch also allows us to combine lexical and semantic search with [hybrid search](https://opensearch.org/docs/latest/search-plugins/hybrid-search/)

Reviewer 2: ```suggestion
It is important to note that lexical search can still be performed on the index, and OpenSearch also allows you to combine lexical and semantic search by using [hybrid search](https://opensearch.org/docs/latest/search-plugins/hybrid-search/).
```

## Summary

> Multilingual embedding models have opened a new door for multilingual search. 

Reviewer 2: ```suggestion
Multilingual embedding models have presented new capabilities for multilingual search. 
```

> Using connectors, connectors blueprints, and ingest pipelines simplifies the ingestion and query process in OpenSearch.

Reviewer 2: ```suggestion
Connectors, connectors blueprints, and ingest pipelines simplify the ingestion and query process in OpenSearch.
```

> There are more considerations to be taken that we have not covered in this blog, such as the quality of the embedding model and the ability to fine tune it, the k-NN engine used, and hybrid search approaches. 

Reviewer 2: ```suggestion
Additional considerations that were not covered in this blog post include the quality of the embedding model and the ability to fine-tune it, the k-NN engine used, and hybrid search approaches. You can find more information about these subjects in the resources provided in the following section.
```

> I have added a section below for further reading.

Reviewer 2: ```suggestion
```

## Further reading
* [k-NN search](https://opensearch.org/docs/latest/search-plugins/knn/index/)
* [Building a semantic search engine in OpenSearch](https://opensearch.org/blog/semantic-search-solutions/)
* [The ABCs of semantic search in OpenSearch: Architectures, benchmarks, and combination strategies](https://opensearch.org/blog/semantic-science-benchmarks/)

> * [Natural Language and Search - Large Language Models (LLMs) for Semantic Search and Generative AI](https://d1.awsstatic.com/aws-analytics-content/OReilly_book_Natural-Language-and-Search_web.pdf)

Reviewer 2: ```suggestion
* [Natural Language and Search: Large Language Models (LLMs) for Semantic Search and Generative AI](https://d1.awsstatic.com/aws-analytics-content/OReilly_book_Natural-Language-and-Search_web.pdf)
```
